homeassistant/components/pvpc_hourly_pricing/sensor.py

```diff
@@ -1,160 +1,105 @@
-            self._pvpc_data.tariff,
-            mins_update,
+            self.coordinator.api.tariff,
         )
-        now = dt_util.utcnow()
-        await self.async_update_prices(now)
-        self.update_current_price(now)
 
-    @property
-    def unique_id(self) -> str | None:
-        """Return a unique ID."""
-        return self._unique_id
-
-    @property
-    def name(self) -> str:
-        """Return the name of the sensor."""
-        return self._name
+    @callback
+    def update_current_price(self, now: datetime) -> None:
+        """Update the sensor state, by selecting the current price for this hour."""
+        self.coordinator.api.process_state_and_attributes(now)
+        self.async_write_ha_state()
 
     @property
-    def native_value(self) -> float:
+    def native_value(self) -> StateType:
         """Return the state of the sensor."""
-        return self._pvpc_data.state
-
-    @property
-    def available(self) -> bool:
-        """Return True if entity is available."""
-        return self._pvpc_data.state_available
+        return self.coordinator.api.state
 
     @property
-    def extra_state_attributes(self) -> dict[str, Any]:
+    def extra_state_attributes(self) -> Mapping[str, Any] | None:
         """Return the state attributes."""
-        return self._pvpc_data.attributes
-
-    @callback
-    def update_current_price(self, now):
-        """Update the sensor state, by selecting the current price for this hour."""
-        self._pvpc_data.process_state_and_attributes(now)
-        self.async_write_ha_state()
-
-    async def async_update_prices(self, now):
-        """Update electricity prices from the ESIOS API."""
-        prices = await self._pvpc_data.async_update_prices(now)
-        if not prices and self._pvpc_data.source_available:
-            self._num_retries += 1
-            if self._num_retries > 2:
-                _LOGGER.warning(
-                    "%s: repeated bad data update, mark component as unavailable source",
-                    self.entity_id,
-                )
-                self._pvpc_data.source_available = False
-                return
-
-            retry_delay = 2 * self._num_retries * self._pvpc_data.timeout
-            _LOGGER.debug(
-                "%s: Bad update[retry:%d], will try again in %d s",
-                self.entity_id,
-                self._num_retries,
-                retry_delay,
-            )
-            async_call_later(self.hass, retry_delay, self.async_update_prices)
-            return
-
-        if not prices:
-            _LOGGER.debug("%s: data source is not yet available", self.entity_id)
-            return
-
-        self._num_retries = 0
-        if not self._pvpc_data.source_available:
-            self._pvpc_data.source_available = True
-            _LOGGER.warning("%s: component has recovered data access", self.entity_id)
-            self.update_current_price(now)
+        return self.coordinator.api.attributes
```

Comment:
Please set the keys and values in a local dict explicitly before we return the dict as extra state attributes. We want to see what the keys are and their formatting and what the values mean.

Reply:
> We want to see what the keys are and their formatting and what the values mean.


The dict with lots of attrs is generated by the library since a long time, the logic was getting big and looked like the best approach to move out all of that and leaving this one lighter. 

Same for the test suite: logic for the download behavior, the DST changes, the tariff periods, and more little nightmares are deeply tested in `aiopvpc` 😅

Also, the keys for this dict are variable 😱. In the evening, the prices for the next day are available, and the dict has more info. In DST changes there is one more/one less hourly price tag, etc.

Logic is in https://github.com/azogue/aiopvpc/blob/600221a5cf82da4b96a7e47f4b8bc32a0fcad8b4/aiopvpc/pvpc_data.py#L245

I have in the TODO list to split some of those in multiple sensors, but didn't want to make breaking changes in this one, is that ok?

Reply:
Please define a dictionary here in the integration with a map of state attribute keys and the keys that may be present in the data returned by the library. Then iterate over this map and set the items that are present in the state attributes.

Reply:
> Please define a dictionary here in the integration with a map of state attribute keys and the keys that may be present in the data returned by the library. Then iterate over this map and set the items that are present in the state attributes.

Hi @MartinHjelmare, that would be something like this 🙈😅:

```python
_KEYS_SENSOR_ATTRIBUTES = (
    "attribution",  # str
    # tariff data
    "tariff",  # str
    "period",  # str
    "available_power",  # int
    "next_period",  # str
    "hours_to_next_period",  # int
    # sensor stats
    "next_better_price",  # float
    "hours_to_better_price",  # int
    "num_better_prices_ahead",  # int
    "price_position",  # int
    "price_ratio",  # float
    "max_price",  # float
    "max_price_at",  # int
    "min_price",  # float
    "min_price_at",  # int
    "next_best_at",  # list[int]
    # hourly prices
    "price_00h",  # float
    "price_01h",  # float
    "price_02h",  # float
    "price_02h_d",  # float, only on DST day
    "price_03h",  # float
    "price_04h",  # float
    "price_05h",  # float
    "price_06h",  # float
    "price_07h",  # float
    "price_08h",  # float
    "price_09h",  # float
    "price_10h",  # float
    "price_11h",  # float
    "price_12h",  # float
    "price_13h",  # float
    "price_14h",  # float
    "price_15h",  # float
    "price_16h",  # float
    "price_17h",  # float
    "price_18h",  # float
    "price_19h",  # float
    "price_20h",  # float
    "price_21h",  # float
    "price_22h",  # float
    "price_23h",  # float
    # only at evening
    "next_better_price (next day)",  # float
    "hours_to_better_price (next day)",  # int
    "num_better_prices_ahead (next day)",  # int
    "price_position (next day)",  # int
    "price_ratio (next day)",  # float
    "max_price (next day)",  # float
    "max_price_at (next day)",  # int
    "min_price (next day)",  # float
    "min_price_at (next day)",  # int
    "next_best_at (next day)",  # list[int]
    "price_next_day_00h",  # float
    "price_next_day_01h",  # float
    "price_next_day_02h",  # float
    "price_next_day_02h_d",  # float, only on DST day
    "price_next_day_03h",  # float
    "price_next_day_04h",  # float
    "price_next_day_05h",  # float
    "price_next_day_06h",  # float
    "price_next_day_07h",  # float
    "price_next_day_08h",  # float
    "price_next_day_09h",  # float
    "price_next_day_10h",  # float
    "price_next_day_11h",  # float
    "price_next_day_12h",  # float
    "price_next_day_13h",  # float
    "price_next_day_14h",  # float
    "price_next_day_15h",  # float
    "price_next_day_16h",  # float
    "price_next_day_17h",  # float
    "price_next_day_18h",  # float
    "price_next_day_19h",  # float
    "price_next_day_20h",  # float
    "price_next_day_21h",  # float
    "price_next_day_22h",  # float
    "price_next_day_23h",  # float
)

self._attrs = {
    key: value
    for key, value in self.coordinator.api.attributes.items()
    if key in _KEYS_SENSOR_ATTRIBUTES
}
```

which doesn't has a lot of sense to my eyes, I don't know if I'm understanding and following you, sorry. The map is not required, as those `api.attributes` come with the desired keys, and the logic to include the keys is in the library, that's why, depending on the moment, the dict has _variable_ keys.

The typing is also kind of complex, but not problematic. The dict is typed as `dict[str, Any]` where the `Any` is really that `float | int | str | list[str]` shown in that `_KEYS_SENSOR_ATTRIBUTES` tuple of keys.

**How should I proceed?** Do you want the code above with the dict comprehension with that always-true if condition?

Reply:
Yes, please. Also, we only want snake_case lowercase keys in the state attributes. So it would be good to make it a dict and normalise the keys in the state attributes like that.

The reason that we want this map is that the integration should be in control of what data is returned as state attributes.

Attribution is set as the corresponding entity property nowadays so we can move that item to that property.

Reply:
Normalizing the keys can be done separately to avoid a breaking change.

Reply:
Ok, understood, thanks @MartinHjelmare  👍

> we only want snake_case lowercase keys in the state attributes

> Normalizing the keys can be done separately to avoid a breaking change

All keys are already normalized except the ones with the `"  (next day)"` suffix 🙈 that appear momentarily in the evening, and those are not in use in the custom lovelace card (https://github.com/danimart1991/pvpc-hourly-pricing-card), so I'll fix those with the new mapping (no need for another PR or even announce a breaking change for them, if you agree)

Reply:
It's still a breaking change if they are changed. I think we should announce a breaking change to make it clear.

Reply:
@MartinHjelmare, a question:

The existence of the breaking change forces to include the PR in the next version `2022.1.0`? or it could be included in the next patch `2021.12.5`?

If the former, then I think I could do the key normalization in other PR, in order to deploy this one as soon as possible, as the integration is broken right now 🥺

Reply:
Yes, do the breaking change in a separate PR. :+1: